README
NAME
    App::AYCABTU - All Your Codes Are Belong To Us

SYNOPSIS
        > aycabtu --help

DESCRIPTION
    This module installs a program called aycabtu, that can be used to
    manage all of the code repositories that you are interested in.

STATUS
    This is a very early release. Only a couple features are implemented,
    and somewhat poorly.

    See "/github.com/ingydotnet/aycabt- ingydotnet/blob/master/AYCABTU" in
    http: for an example of how to configure AYCABTU.

    Stay tuned. Things should be much better soon.

RESOURCES
    CPAN: <http://search.cpan.org/dist/App-AYCABTU/>

    GitHub: <http://github.com/ingydotnet/app-aycabtu-pm>

AUTHOR
    Ingy döt Net <ingy@cpan.org>

COPYRIGHT
    Copyright (c) 2010. Ingy döt Net

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://www.perl.com/perl/misc/Artistic.html>



Changes
---
version: 0.13
date:    Sat Aug 16 15:46:02 PDT 2014
changes:
- Remove versions from module use
---
version: 0.12
date:    Sat Aug 16 13:19:52 PDT 2014
changes:
- Some doc fixes. mst++ for reals this time.
---
version: 0.11
date:    Sat Aug 16 13:14:27 PDT 2014
changes:
- Fix Meta homepage. mst++
---
version: 0.10
date:    Sat Aug 16 10:47:40 PDT 2014
changes:
- Meta 0.0.2
---
version: 0.09
date:    Fri Aug 15 18:02:47 PDT 2014
changes:
- Remove t/_
---
version: 0.08
date:    Fri Aug 15 15:39:33 PDT 2014
changes:
- Add t/000-compile-modules.t
- CPAN Day 2014 - 21 Release Salute!
---
version: 0.07
date:    Fri Aug  8 11:50:03 PDT 2014
changes:
- Swith hashbangs to `/usr/bin/env perl`
---
version: 0.06
date:    Sat Aug  2 10:03:19 PDT 2014
changes:
- Fix Copyright years.
---
version: 0.05
date:    Sat Jul 26 22:04:46 PDT 2014
changes:
- Switch to Zilla-Dist
---
version: 0.04
date:    Mon Jun  6 00:15:51 EST 2011
changes:
- Allow single numbers for selection
- Switch to Module::Package and Mouse.
---
version: 0.03
date:    Thu Nov 18 22:31:39 PST 2010
changes:
- Glob together multiple config files.
---
version: 0.02
date:    Tue Oct  5 21:30:41 PDT 2010
changes:
- Add git clone support
- Add names selection
---
version: 0.01
date:    Sat Sep 18 00:36:35 PDT 2010
changes:
- Initial version.


lib/App/AYCABTU/OO.pm
### This module was derived from Gloom - the Great Little OO Module!
### Read `perldoc Gloom` for more information.

package App::AYCABTU::OO;

use 5.008003;
use strict;
use warnings;

our $VERSION = '0.15';

sub import {
    my ($class, $flag) = @_;
    my ($package, $module) = caller(0);

    strict->import;
    warnings->import;

    if ($class->isa(__PACKAGE__) and
        defined $flag and
        $flag eq '-base'
    ) {
        $class->import_base($package, $module);
    }
    else {
        require Exporter;
        goto &Exporter::import;
    }
}

sub import_base {
    my ($class, $package, $module) = @_;
    no strict 'refs';
    push @{$package . '::ISA'}, $class;
    $class->import_fake($package, $module);
    $class->export_base($package);
}

sub import_fake {
    my ($class, $package, $module) = @_;
    my $inc_module = $package . '.pm';
    $inc_module =~ s/::/\//g;
    return if defined $INC{$inc_module};
    $INC{$inc_module} = $module;
}

sub export_base {
    my ($source, $target) = @_;
    no strict 'refs';
    for my $sub (map {
        /::/ ? $_ : "${source}::$_"
    } $source->EXPORT_BASE()) {
        my $name = $sub;
        $name =~ s/.*:://;
        *{$target . "::$name"} = \&$sub;
    }
}

sub new {
    my $class = shift;
    my $self = bless {}, $class;
    $self->init(@_);
    return $self;
}

sub init {
    my $self = shift;
    while (my ($property, $value) = splice(@_, 0, 2)) {
        unless ($self->can($property)) {
            my $class = ref $self;
            require Carp;
            Carp::confess("Class '$class' has no property '$property'");
        }
        $self->$property($value);
    }
}

my %code = (
    sub_start =>
      "sub {\n",
    set_default =>
      "  \$_[0]->{%s} = %s\n    unless exists \$_[0]->{%s};\n",
    class =>
      "  return do { my \$class = \$_[0]; %s } unless ref \$_[0];\n",
    init =>
      "  return \$_[0]->{%s} = do { my \$self = \$_[0]; %s }\n" .
      "    unless \$#_ > 0 or defined \$_[0]->{%s};\n",
    return_if_get =>
      "  return \$_[0]->{%s} unless \$#_ > 0;\n",
    set =>
      "  \$_[0]->{%s} = \$_[1];\n",
    onset =>
      "  do { local \$_ = \$_[1]; my \$self = \$_[0]; %s };\n",
    chain =>
      "  return \$_[0];\n}\n",
    sub_end => 
      "  return \$_[0]->{%s};\n}\n",
);

my $parse_arguments = sub {
    my $paired_arguments = shift || []; 
    my ($args, @values) = ({}, ());
    my %pairs = map { ($_, 1) } @$paired_arguments;
    while (@_) {
        my $elem = shift;
        if (defined $elem and defined $pairs{$elem} and @_) {
            $args->{$elem} = shift;
        }
        elsif ($elem eq '-chain') {
            $args->{-chain} = 1;
        }
        else {
            push @values, $elem;
        }
    }
    return wantarray ? ($args, @values) : $args;        
};

my $default_as_code = sub {
    no warnings 'once';
    require Data::Dumper;
    local $Data::Dumper::Sortkeys = 1;
    my $code = Data::Dumper::Dumper(shift);
    $code =~ s/^\$VAR1 = //;
    $code =~ s/;$//;
    return $code;
};

sub has {
    my $package = caller;
    my ($args, @values) = &$parse_arguments(
        [ qw(-package -class -init -onset) ],
        @_,
    );
    my ($has, $default) = @values;
    $package = $args->{-package} if defined $args->{-package};
    return if defined &{"${package}::$has"};
    my $default_string =
        ( ref($default) eq 'ARRAY' and not @$default )
        ? '[]'
        : (ref($default) eq 'HASH' and not keys %$default )
          ? '{}'
          : &$default_as_code($default);

    my $code = $code{sub_start};

    if ($args->{-class}) {
        if ($args->{-class} eq '-init') {
            $args->{-class} = $args->{-init};
            $args->{-class} =~ s/\$self/\$class/g;
        }
        my $fragment = $code{class};
        $code .= sprintf
            $fragment,
            $args->{-class};
    }

    if ($args->{-init}) {
        my $fragment = $code{init};
        $code .= sprintf
            $fragment,
            $has,
            $args->{-init},
            ($has) x 4;
    }
    $code .= sprintf $code{set_default}, $has, $default_string, $has
      if defined $default;
    $code .= sprintf $code{return_if_get}, $has;
    $code .= sprintf $code{set}, $has;
    $code .= sprintf $code{onset}, $args->{-onset}
      if defined $args->{-onset};
    if (defined $args->{-chain}) {
        $code .= $code{chain};
    }
    else {
        $code .= sprintf $code{sub_end}, $has;
    }

    my $sub = eval $code;
    die "eval('$code') failed: $@" if $@;
    no strict 'refs';
    *{"${package}::$has"} = $sub;
    return $code if defined wantarray;
}

our $DumpModule = 'YAML::XS';
sub WWW { require XXX; local $XXX::DumpModule = $DumpModule; XXX::WWW(@_) }
sub XXX { require XXX; local $XXX::DumpModule = $DumpModule; XXX::XXX(@_) }
sub YYY { require XXX; local $XXX::DumpModule = $DumpModule; XXX::YYY(@_) }
sub ZZZ { require XXX; local $XXX::DumpModule = $DumpModule; XXX::ZZZ(@_) }

sub EXPORT_BASE {
    return map { __PACKAGE__ . '::' .$_ }
        qw(has WWW XXX YYY ZZZ);
}


lib/App/AYCABTU.pm
package App::AYCABTU;
our $VERSION = '0.12';

use Mouse 0.93;

use Getopt::Long;
use YAML::XS 0.35;
use Capture::Tiny 0.10 'capture';

has config => (is => 'ro', default => sub{[]});

has file => (is => 'ro', default => 'AYCABTU');
has action => (is => 'ro', default => 'list');
has show => (is => 'ro', default => '');
has tags => (is => 'ro', default => sub{[]});
has names => (is => 'ro', default => sub{[]});
has all => (is => 'ro', default => 0);
has quiet => (is => 'ro', default => 0);
has verbose => (is => 'ro', default => 0);
has args => (is => 'ro', default => sub{[]});

has repos => (is => 'ro', default => sub{[]});

my ($prefix, $error, $quiet, $normal, $verbose);

sub run {
    my $self = shift;
    my @opts = @_
        ? @_
        : split /\s+/, ($ENV{AYCABTU_DEFAULT_OPTS} || '');
    $self->get_options(@opts);
    $self->read_config();
    $self->select_repos();
    if (not @{$self->repos} and not @{$self->names}) {
        print "No repositories selected. Try --all.\n";
        return;
    }
    my $action = $self->action;
    my $method = "action_$action";
    die "Can't perform action '$action'\n"
        unless $self->can($method);
    for my $entry (@{$self->repos}) {
        ($prefix, $error, $quiet, $normal, $verbose) = ('') x 5;
        $self->$method($entry);
        $verbose ||= $normal;
        $normal ||= $quiet;
        my $msg =
            $error ? $error :
            $self->verbose ? $verbose :
            $self->quiet ? $quiet :
            $normal;
        $msg = "$prefix$msg\n" if $msg;
        print $msg;
    }
    if (@{$self->names}) {
        warn "The following names were not found: @{$self->names}\n";
    }
}

sub get_options {
    my $self = shift;
    local @ARGV = @_;
    GetOptions(
        'file=s' => sub { $self->file($_[1]) },
        'verbose' => sub { $self->verbose(1) },
        'quiet' => sub { $self->quiet(1) },
        'list' => sub { $self->action('list') },
        'update' => sub { $self->action('update') },
        'status' => sub { $self->action('status') },
        'show=s' => sub { $self->action('show'); $self->show($_[1]) },
        'all' => sub { $self->all(1) },
        'tags=s' => sub {
            my $tags = $_[1] or return;
            push @{$self->tags}, [split ',', $tags];
        },
        'help' => \&help,
    );
    no warnings;
    my $names;
    if (1 or not -t stdin) {
        $names = [
            map {
                s!/$!!;
                /^(\d+)-(\d+)?$/ ? ($1..$2) :
                /^(\d+)$/ ? ($1) :
                (-d) ? ($_) :
                ();
            } @ARGV
        ];
    }
    else {
        $names = [ split /\s+/, do {local $/; <stdin>} ]
    }
    $self->names($names);
    die "Can't locate aycabtu config file '${\ $self->file}'. Use --file=... option\n"
        if not -e $self->file and not @{[glob $self->file . '*']};
}

sub read_yaml {
    my $self = shift;
    my @files = glob($self->file . '*');
    my $yaml = '';
    local $/;
    for my $file (@files) {
        open Y, $file;
        $yaml .= <Y>;
    }
    return $yaml;
}

sub read_config {
    my $self = shift;
    my $yaml = $self->read_yaml();
    my $config = YAML::XS::Load($yaml);
    $self->config($config);
    die $self->file . " must be a YAML sequence of mapping"
        if (ref($config) ne 'ARRAY') or grep {
            ref ne 'HASH'
        } @$config;
    my $count = 1;
    for my $entry (@$config) {
        my $repo = $entry->{repo}
            or die "No 'repo' field for entry $count";

        $entry->{_num} = $count++;

        $entry->{name} ||= '';
        if (not $entry->{name} and $repo =~ /.*\/(.*).git$/) {
            my $name = $1;
            # XXX This should be configable.
            $name =~ s/\.wiki$/-wiki/;
            $entry->{name} = $name;
        }

        my $type = $entry->{type}  || '';
        $type ||=
            ($repo =~ /\.git$/) ? 'git' :
            ($repo =~ /svn/) ? 'svn' :
            '';
        $entry->{type} = $type;

        my $tags = $entry->{tags} || '';

        my $set = $tags ? { map {($_, 1)} split /[\s\,]+/, $tags } : {};
        my $str = $repo;
        $str =~ s/\/$//;
        $str =~ s/\/trunk$//;
        $str =~ s/.*\///;
        my $subst = {
            py => 'python',
            pm => 'perl',
        };
        $set->{$_} = 1 for map {$subst->{$_} || $_} split /[^\w]+/, $str;
        $set->{$type} = 1;
        delete $set->{''};

        $entry->{tags} = [ sort map lc, keys %$set ];
    }
}

sub select_repos {
    my $self = shift;

    my $config = $self->config;
    my $repos = $self->repos;
    my $names = $self->names;

    my $last = 0;
OUTER:
    for my $entry (@$config) {
        last if $last;
        next if $entry->{skip};
        $last = 1 if $entry->{last};

        if ($self->all) {
            push @$repos, $entry;
            next;
        }
        my ($num, $name) = @{$entry}{qw(_num name)};
        if (@$names) {
            if (grep {$_ eq $name or $_ eq $num} @$names) {
                push @$repos, $entry;
                @$names = grep {$_ !~ /^(\Q$name\E|$num)$/} @$names;
                next;
            }
        }
        for my $tags (@{$self->tags}) {
            if ($tags) {
                my $count = scalar grep {
                    my $t = $_;
                    grep {$_ eq $t} @{$entry->{tags}};
                } @$tags;
                if ($count == @$tags) {
                    push @$repos, $entry;
                    next OUTER;
                }
            }
        }
    }
}

sub action_update {
    my $self = shift;
    my $entry = shift;
    $self->_check(update => $entry) or return;
    my ($num, $name) = @{$entry}{qw(_num name)};
    $prefix = "$num) Updating $name... ";
    $self->git_update($entry);
}

sub action_status {
    my $self = shift;
    my $entry = shift;
    $self->_check('check status' => $entry) or return;
    my ($num, $name) = @{$entry}{qw(_num name)};
    $prefix = "$num) Status for $name... ";
    $self->git_status($entry);
}

sub action_list {
    my $self = shift;
    my $entry = shift;
    my ($num, $repo, $name, $type, $tags) = @{$entry}{qw(_num repo name type tags)};
    $prefix = "$num) ";
    $quiet = $name;
    $normal = sprintf " %-25s %-4s %-50s", $name, $type, $repo;
    $verbose = "$normal\n    tags: @$tags";
}

sub action_show {
    my $self = shift;
    my $entry = shift;
    my $show = $self->show;
    $prefix = '';
    if ($show =~ /^(nums?|numbers?)$/) {
        $quiet = $entry->{_num};
    }
    elsif ($show =~ /^names?$/) {
        $quiet = $entry->{name};
    }
    elsif ($show =~ /^tags?$/) {
        my $set = {};
        for my $repo (@{$self->repos}) {
            $set->{$_} = 1 for @{$repo->{tags}};
        }
        my @tags = sort keys %$set;
        print "@tags\n";
        exit;
    }
    else {
        $error = "Invalid type '$show' to show.";
    }
}

sub _check {
    my $self = shift;
    my $action = shift;
    my $entry = shift;
    my ($num, $repo, $name, $type) = @{$entry}{qw(_num repo name type)};
    if (not $name) {
        $error = "Can't $action $repo. No name.";
        return;
    }
    if (not $type) {
        $error = "Can't $action $name. Unknown type.";
        return;
    }
    if ($type ne 'git') {
        $error = "Can't $action $name. Type $type not yet supported.";
        return;
    }
    return 1;
}

sub git_update {
    my $self = shift;
    my $entry = shift;
    my ($repo, $name) = @{$entry}{qw(repo name)};
    if (not -d $name) {
        my $cmd = "git clone $repo $name";
        my ($o, $e) = capture { system($cmd) };
        if ($e =~ /\S/) {
            $quiet = 'Error';
            $verbose = "\n$o$e";
        }
        else {
            $normal = 'Done';
        }
    }
    elsif (-d "$name/.git") {
        my ($o, $e) = capture { system("cd $name; git pull origin master") };
        if ($o eq "Already up-to-date.\n") {
            $normal = "Already up to date";
        }
        elsif ($e) {
            $quiet = "Failed";
            $verbose = "\n$o$e";
        }
        else {
            $quiet = "Updated";
            $verbose = "\n$o$e";
        }
    }
    else {
        $quiet = "Skipped";
    }
}

sub git_status {
    my $self = shift;
    my $entry = shift;
    my ($repo, $name) = @{$entry}{qw(repo name)};
    if (not -d $name) {
        $error = "No local repository";
    }
    elsif (-d "$name/.git") {
        my ($o, $e) = capture { system("cd $name; git status") };
        if ($o =~ /^nothing to commit/m and
            not $e
        ) {
            if ($o =~ /Your branch is ahead .* by (\d+) /) {
                $quiet = "Ahead by $1";
                $verbose = "\n$o$e";
            }
            else {
                $normal = "OK";
            }
        }
        else {
            $quiet = "Dirty";
            $verbose = "\n$o$e";
        }
    }
    else {
        $quiet= "Skipped";
    }
}

sub help {
    print <<'...';
Usage:
    aycabtu [ options ] action selectors

Options:
    --file=file     # aycabtu config file. Default: 'AYCABTU'
    --verbose       # Show more information
    --quiet         # Show less information

Action:
    --list          # List the selected repos (default action)
    --update        # Checkout or update the selected repos
    --status        # Get status info on the selected repos
    --show=aspect   # Show some aspect of the selected repos

Show Aspects:
    numbers         # Show the numbers of the selected repos
    names           # Show the numbers of the selected repos
    tags            # Show ALL tags of selected repos

Selector:
    --all           # Use all the repos in the config file
    --tags=tags     # Select repos matching all the tags
                      Can be used more than once
    names           # A list of the names to to select. You can use
                    # multiple names and file globbing, like this:

        aycabtu --update foo-repo bar-*-repo

...
    exit;
}

1;
